lib/testutils/mini_runner.go

```diff
@@ -0,0 +1,137 @@
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package testutils
+
+import (
+	"context"
+	"sync/atomic"
+
+	"github.com/loadimpact/k6/lib"
+	"github.com/loadimpact/k6/stats"
+)
+
+// Ensure mock implementations conform to the interfaces.
+var _ lib.Runner = &MiniRunner{}
+var _ lib.VU = &MiniRunnerVU{}
+
+// MiniRunner partially implements the lib.Runner interface, but instead of
+// using a real JS runtime, it allows us to directly specify the options and
+// functions with Go code.
+type MiniRunner struct {
+	Fn         func(ctx context.Context, out chan<- stats.SampleContainer) error
+	SetupFn    func(ctx context.Context, out chan<- stats.SampleContainer) ([]byte, error)
+	TeardownFn func(ctx context.Context, out chan<- stats.SampleContainer) error
+
+	setupData []byte
+
+	CurrentVU int64
+	Group     *lib.Group
+	Options   lib.Options
+}
+
+// MakeArchive isn't implemented, it always returns nil and is just here to
+// satisfy the lib.Runner interface.
+func (r MiniRunner) MakeArchive() *lib.Archive {
+	return nil
+}
+
+// NewVU returns a new MiniRunnerVU with an incremental ID.
+func (r *MiniRunner) NewVU(out chan<- stats.SampleContainer) (lib.VU, error) {
+	nextVUNum := atomic.AddInt64(&r.CurrentVU, 1)
+	return &MiniRunnerVU{R: r, Out: out, ID: nextVUNum - 1}, nil
+}
+
+// Setup just calls the supplied mock setup() function, if present.
+func (r *MiniRunner) Setup(ctx context.Context, out chan<- stats.SampleContainer) (err error) {
+	if fn := r.SetupFn; fn != nil {
+		r.setupData, err = fn(ctx, out)
+	}
+	return
+}
+
+// GetSetupData returns json representation of the setup data if setup() is
+// specified and was ran, nil otherwise.
+func (r MiniRunner) GetSetupData() []byte {
+	return r.setupData
+}
+
+// SetSetupData saves the externally supplied setup data as JSON in the runner.
+func (r *MiniRunner) SetSetupData(data []byte) {
+	r.setupData = data
+}
+
+// Teardown just calls the supplied mock teardown() function, if present.
+func (r MiniRunner) Teardown(ctx context.Context, out chan<- stats.SampleContainer) error {
+	if fn := r.TeardownFn; fn != nil {
+		return fn(ctx, out)
+	}
+	return nil
+}
+
+// GetDefaultGroup just returns the default group.
+func (r MiniRunner) GetDefaultGroup() *lib.Group {
+	if r.Group == nil {
+		r.Group = &lib.Group{}
+	}
+	return r.Group
+}
+
+// GetOptions just returns the supplied options struct.
```

Comment:
Nit: as mentioned in #1007, consider dropping the word "just" from struct/method comments as it trivializes the implementation and doesn't add much value. In this case "GetOptions returns the supplied options struct" is a much clearer and concise description IMHO.

Reply:
Resolved by the [latest force-push](https://github.com/loadimpact/k6/compare/8b97f2b5ec20e5996f35eb8791d5adacfc095204..f935df854229e4d8a121aac03d6a8e43eef5ad80)